engine/src/state_chain_observer/client.rs

```diff
@@ -795,52 +806,84 @@ impl SignedExtrinsicClientBuilderTrait for SignedExtrinsicClientBuilder {
 					recorded_version, *CFE_VERSION
 				);
 
-				let block_hash = finalized_block_stream.cache().hash;
-				let block_number = finalized_block_stream.cache().number;
-
-				// Submitting transaction with subxt sometimes gets stuck without returning any
-				// error (see https://linear.app/chainflip/issue/PRO-1064/new-cfe-version-gets-stuck-on-startup),
-				// so we use a timeout to ensure we can recover:
-				tokio::time::timeout(CFE_VERSION_SUBMIT_TIMEOUT, async {
-					let current_nonce = rpc_client
-						.request::<u32>(
-							"system_accountNextIndex",
-							subxt::rpc_params![&subxt_signer.account_id()],
-						)
-						.await?;
+				// Submitting the CFE version has three possible results:
+				//   - Successful
+				//   - Returns no error, but not successful
+				//   - Returns error
+				// Here we call the submission code multiple times before giving up with an error.
+				const MAX_UPDATE_VERSION_RETRIES: usize = 10;
+				let mut update_successful = false;
+				for retry in 0..MAX_UPDATE_VERSION_RETRIES {
+					let block_hash = finalized_block_stream.cache().hash;
+					let block_number = finalized_block_stream.cache().number;
+
+					// Submitting transaction with subxt sometimes gets stuck without returning any
+					// error (see https://linear.app/chainflip/issue/PRO-1064/new-cfe-version-gets-stuck-on-startup),
+					// so we use a timeout to ensure we can recover:
+					let result = tokio::time::timeout(CFE_VERSION_SUBMIT_TIMEOUT, async {
+						let current_nonce = rpc_client
+							.request::<u32>(
+								"system_accountNextIndex",
+								subxt::rpc_params![&subxt_signer.account_id()],
+							)
+							.await?;
+
+						subxt_client
+							.tx()
+							.create_signed_with_nonce(
+								&subxt::dynamic::tx(
+									"Validator",
+									"cfe_version",
+									vec![(
+										"new_version",
+										vec![
+											("major", CFE_VERSION.major),
+											("minor", CFE_VERSION.minor),
+											("patch", CFE_VERSION.patch),
+										],
+									)],
+								),
+								&subxt_signer,
+								current_nonce.into(),
+								DefaultExtrinsicParamsBuilder::new()
+									.mortal_unchecked(
+										block_number.into(),
+										block_hash,
+										SIGNED_EXTRINSIC_LIFETIME.into(),
+									)
+									.build(),
+							)?
+							.submit_and_watch()
+							.await?
+							.wait_for_finalized()
+							.await
+					})
+					.await
+					.map_err(|_| anyhow::anyhow!("Timed out trying to submit CFE version"))
+					.map(|x| x.map_err(|e| e.into())); // map inner subxt::Error to anyhow::Error
+
+					match result.flatten() {
+						Ok(_) => {
+							// Submitting did not return an error, but this does not mean it was
+							// successful. Thus we check whether the version was actually updated:
+							let new_version =
+								get_recorded_cfe_version(&subxt_client, &subxt_signer).await?;
+							if new_version == *CFE_VERSION {
```

Comment:
nit:
it might not actually be the "new version" in the case of it not updating, can just be inlined to avoid naming it
```suggestion
							if get_recorded_cfe_version(&subxt_client, &subxt_signer).await? == *CFE_VERSION {
```